app/models/concerns/doi_generator.rb

```diff
@@ -0,0 +1,170 @@
+module DoiGenerator
+  extend ActiveSupport::Concern
+
+  attr_accessor :doi_message
+
+  def check_doi_presence
+    return self.doi_message unless can_get_doi?
+    if self.doi.present?
+      update_datacite_doi
+    else
+      create_datacite_doi
+    end
+    self.doi_message
+  end
+
+  private
+
+  def can_get_doi?
+    # Only generate if required metadata is there
+    unless self.id.present? && self.creator.present? && self.title.present?
+      self.doi_message = 'metadata'
+      return false
+    end
+    # Only generate for combined pages
+    if self.is_a?(Page) && self.page_number.present?
+      self.doi_message = 'page'
+      return false
+    end
+    true
+  end
+
+  def create_datacite_doi
+    response = DataciteRest.new.mint(
+      datacite_api_json
+    )
+    data = JSON[response.body]["data"]
+    self.update_attributes(doi: [data["id"]])
+    self.doi_message = 'generated'
+    self.doi_message = 'generated_draft' if data["attributes"]["state"] == 'draft'
+  end
+
+  def update_datacite_doi
+    client = DataciteRest.new
+
+    self.doi.each do |doi|
+      begin
+        response = client.get_doi(doi)
+        data = JSON[response.body]["data"]
+        identifier = data["id"]
+        new_state = visibility_to_state
+        current_state = data["attributes"]["state"]
+        # if state changes from findable to anything else, we hide it
+        if current_state == "findable" && new_state != "findable"
+          json = datacite_api_json(hide=true)
+        else
+          json = datacite_api_json
+        end
+        client.update_metadata(identifier, json)
+        update_doi_metadata_message(current_state, new_state)
```

Comment:
It's great that a separate Datacite library wasn't needed in the end.

There might be state transition issues, so making the state transitions explicit in the code may help identify them. In any case, there is a state machine here by any other name in that code:

![visibility_doi_state_machine](https://user-images.githubusercontent.com/1932651/148255867-b4865e3e-17dd-44ac-b51f-45413a05374c.png)

So something like the following to make it explicit:


```ruby
        response = client.get_doi(doi)
        data = JSON[response.body]["data"]
        identifier = data["id"]
        doi_state = data["attributes"]["state"]
        action, message = get_action_message(doi_state, self.visibility)
        json = datacite_api_json(action)
        client.update_metadata(identifier, json)
        update_doi_metadata_message(message)
```


```ruby
def get_action_message(src_doi_state, target_visiblity)
  states = {
    # (<doi_state>, <visibility>): (<action>, <message>)
  }
  return states[(src_doi_state, target_visibility)]  # add any weird exceptions here or in the states perhaps
end
```

You can then get rid of `set_datacite_event`, `visibility_to_state`....

Test what happens when a unsupported transition on Datacite REST API is executed and adjust for that. (e.g. Datacite says that "registered" or "findable" DOIs can never transition back to "draft").

Reply:
Thank you for this comment, Guillaume, this was helpful. I am glad you noticed the state machine. I knew I needed another pair of eyes after needing to just get this to work. I removed `set_datacite_event` and `visibility_to_state`, hopefully making the transitions easier to read.